apollo-client/src/main/java/com/ctrip/framework/apollo/monitor/internal/listener/AbstractApolloClientMonitorEventListener.java

```diff
@@ -0,0 +1,141 @@
+import com.ctrip.framework.apollo.monitor.internal.model.CounterModel;
+import com.ctrip.framework.apollo.monitor.internal.model.GaugeModel;
+import com.ctrip.framework.apollo.monitor.internal.model.SampleModel;
+import com.google.common.collect.Maps;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * @author Rawven
+ */
+public abstract class AbstractApolloClientMonitorEventListener implements
+    ApolloClientMonitorEventListener {
+
+  private final Map<String, CounterModel> counterSamples = Maps.newConcurrentMap();
+  private final Map<String, GaugeModel> gaugeSamples = Maps.newConcurrentMap();
+  private final AtomicBoolean isUpdated = new AtomicBoolean();
+  private final String tag;
+
+  public AbstractApolloClientMonitorEventListener(String tag) {
+    this.tag = tag;
+  }
+
+  @Override
+  public String getName() {
+    return tag;
+  }
+
+  @Override
+  public boolean isSupport(ApolloClientMonitorEvent event) {
+    return tag.equals(event.getTag());
+  }
+
+  @Override
+  public void collect(ApolloClientMonitorEvent event) {
+    collect0(event);
+    isUpdated.set(true);
+  }
+
+  @Override
+  public boolean isMetricsSampleUpdated() {
+    return isUpdated.getAndSet(false);
+  }
+
+  @Override
+  public List<SampleModel> export() {
+    export0();
+    List<SampleModel> samples = new ArrayList<>(counterSamples.values());
+    samples.addAll(gaugeSamples.values());
+    return samples;
+  }
+
+  /**
+   * Specific collection logic
+   */
+  protected void collect0(ApolloClientMonitorEvent event) {
+  }
+
+  /**
+   * Convenient for indicators that can only be obtained from the status object
+   */
+  protected void export0() {
+  }
+
+
+  /**
+   * tool method for updating indicator model
+   */
+  public void createOrUpdateGaugeSample(String metricsName, String[] tagKeys, String[] tagValues,
+      double value) {
+    createOrUpdateSample(metricsName, tagKeys, tagValues, value, false);
+  }
+
+  public void createOrUpdateGaugeSample(String metricsName, double value) {
+    createOrUpdateSample(metricsName, null, null, value, false);
+  }
+
+  public void createOrUpdateCounterSample(String metricsName, String[] tagKeys, String[] tagValues,
+      double increaseValue) {
+    createOrUpdateSample(metricsName, tagKeys, tagValues, increaseValue, true);
+  }
+
+  public void createOrUpdateCounterSample(String metricsName, double increaseValue) {
+    createOrUpdateSample(metricsName, null, null, increaseValue, true);
+  }
+
+  private void createOrUpdateSample(String metricsName, String[] tagKeys, String[] tagValues,
+      double value, boolean isCounter) {
+    String mapKey = metricsName + (tagValues != null ? Arrays.toString(tagValues) : "");
+
+    if (isCounter) {
+      if (!counterSamples.containsKey(mapKey)) {
+        counterSamples.put(mapKey,
```

Comment:
There is thread-safe problem here, we can use map's method computeIfAbsent to solve it

----

Should check all logic about `if (xxx not exists) then put`